core-bundle/src/Cron/Cron.php

```diff
@@ -135,11 +137,39 @@ private function doRun(array $cronJobs, string $scope, bool $force = false): voi
             $repository->unlockTable();
         }
 
-        // Execute all cron jobs to be run
-        foreach ($cronJobsToBeRun as $cron) {
+        $this->executeCrons($cronJobsToBeRun, $scope);
+    }
+
+    /**
+     * @param array<CronJob> $crons
+     */
+    private function executeCrons(array $crons, string $scope): void
+    {
+        /** @var array<string, PromiseInterface> $promises */
+        $promises = [];
+
+        foreach ($crons as $cron) {
             $this->logger?->debug(sprintf('Executing cron job "%s"', $cron->getName()));
 
-            $cron($scope);
+            $return = $cron($scope);
+
+            if ($return instanceof PromiseInterface) {
+                $this->logger?->debug(sprintf('Cron job "%s" is asynchronous. Adding to waiting queue', $cron->getName()));
+
+                $promises[$cron->getName()] = $return;
+            }
+        }
+
+        if (0 === \count($promises)) {
+            return;
+        }
+
+        foreach (Utils::settle($promises)->wait() as $cronjobName => $result) {
```

Comment:
This would be executed when all promises are fulfilled, no? Wouldn't we want to log immediately after resolving each promise, i.e. by chaining one to `$return` a few lines up?

Reply:
Right, adjusted in https://github.com/contao/contao/pull/5554/commits/bb16ced847221d1a9b1e53f85eba002d9a261ceb.